lib/SlackBot.js

```diff
@@ -225,113 +225,253 @@ function Slackbot(configuration) {
+            }
+
+            // set up a couple of special cases based on subtype
+            if (message.subtype && message.subtype == 'channel_join') {
+                // someone joined. maybe do something?
+                if (message.user == bot.identity.id) {
+                    message.type = 'bot_channel_join';
+                } else {
+                    message.type = 'user_channel_join';
+                }
+            } else if (message.subtype && message.subtype == 'group_join') {
+                // someone joined. maybe do something?
+                if (message.user == bot.identity.id) {
+                    message.type = 'bot_group_join';
+                } else {
+                    message.type = 'user_group_join';
+                }
+            } else if (message.subtype) {
+                message.type = message.subtype;
+            } else if (message.channel.match(/^D/)) {
+                // this is a direct message
+                if (message.user == bot.identity.id && message.bot_id) {
+                    return false;
+                }
+                if (!message.text) {
+                    // message without text is probably an edit
+                    return false;
+                }
+
+                // remove direct mention so the handler doesn't have to deal with it
+                message.text = message.text.replace(direct_mention, '')
+                .replace(/^\s+/, '').replace(/^\:\s+/, '').replace(/^\s+/, '');
+
+                message.type = 'direct_message';
+
+            } else {
+                if (message.user == bot.identity.id && message.bot_id) {
+                    return false;
+                }
+                if (!message.text) {
+                    // message without text is probably an edit
+                    return false;
+                }
+
+                if (message.text.match(direct_mention)) {
+                    // this is a direct mention
+                    message.text = message.text.replace(direct_mention, '')
+                    .replace(/^\s+/, '').replace(/^\:\s+/, '').replace(/^\s+/, '');
+                    message.type = 'direct_mention';
+
+                } else if (message.text.match(mention)) {
+                    message.type = 'mention';
+                } else {
+                    message.type = 'ambient';
+                }
+            }
+        }
+
+        // move on to the next stage of the pipeline
+        next();
+
+    });
+
+
     /* Handler functions for the various ways Slack might send a message to
      * Botkit via webhooks.  These include interactive messages (button clicks),
      * events api (messages sent over web hook), slash commands, and outgoing webhooks
      * (patterns matched in slack that result in a webhook)
      */
-    function handleInteractiveMessage(payload, bot) {
+    slack_botkit.middleware.normalize.use(function handleInteractiveMessage(bot, message, next) {
```

Comment:
I like how different message types are different middleware. I think it would be nice to externalize these handlers into separate files, that would reduce the length of this file and make everything easier to read. I'm thinking something like `lib/slack/middleware`